domain/access/state/user.go

```diff
@@ -723,6 +698,118 @@ VALUES ($M.uuid, false)
 	return nil
 }
 
+// UpdateLastModelLogin updates the last login time for the user
+// with the supplied uuid on the model with the supplied model uuid.
+// The following error types are possible from this function:
+// - accesserrors.UserNameNotValid: When the username is not valid.
+// - accesserrors.UserNotFound: When the user cannot be found.
+// - modelerrors.NotFound: If no model by the given modelUUID exists.
+func (st *UserState) UpdateLastModelLogin(ctx context.Context, name string, modelUUID coremodel.UUID) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting DB access")
+	}
+
+	uuidStmt, err := st.getActiveUUIDStmt()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertModelLoginStmt, err := st.Prepare(`
+INSERT INTO model_last_login (model_uuid, user_uuid, time)
+-- The strftime formatter below inserts the time with millisecond precision.
+-- This is useful for testing.
+VALUES ($dbModelAccess.model_uuid, $dbModelAccess.user_uuid, strftime('%Y-%m-%d %H:%M:%f', 'now'))
+ON CONFLICT(model_uuid, user_uuid) DO
+UPDATE SET time = excluded.time`, dbModelAccess{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert model login query")
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		userUUID, err := st.uuidForName(ctx, tx, uuidStmt, name)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		ma := dbModelAccess{
+			UserUUID:  userUUID.String(),
+			ModelUUID: modelUUID.String(),
+		}
+		if err := tx.Query(ctx, insertModelLoginStmt, ma).Run(); err != nil {
+			if internaldatabase.IsErrConstraintForeignKey(err) {
+				// The foreign key constrain may be triggered if the user or the
+				// model does not exist. However, the user must exist or the
+				// uuidForName query would have failed, so it must be the model.
+				return modelerrors.NotFound
```

Comment:
Can we please add more context to this error?

`fmt.Errorf("updating last model login for user %q on model %q: %w")`

Reply:
Added.